File: cs/Problems/Day4.cs
namespace aoc24.Problems;

// https://adventofcode.com/2024/day/4
public sealed class Day4 : IProblem<int>
{
    public int Solve(string input) => CountXmasOccurrencesOptimized(input);

    private static int CountXmasOccurrencesOptimized(ReadOnlySpan<char> input)
    {
        Span<char> inner = stackalloc char[Matrix<char>.SizeFor(input)];
        var matrix = Matrix<char>.CreateFrom(input, inner);

        int occurrences = 0;

        for (int x = 0; x < matrix.Width; x++)
        {
            for (int y = 0; y < matrix.Height; y++)
            {
                Point point = new(x, y);
                occurrences += FindWordMatches(matrix, point, "XMAS");
            }
        }

        return occurrences;
    }

    private static int CountXmasOccurrences(string input)
    {
        // The only difference is that this version of "CreateFrom" will allocate a backing array for the matrix,
        // as opposed to the other one which will use the provided and stack allocated structure...
        var matrix = Matrix<char>.CreateFrom(input);

        int occurrences = 0;

        for (int x = 0; x < matrix.Width; x++)
        {
            for (int y = 0; y < matrix.Height; y++)
            {
                Point point = new(x, y);
                occurrences += FindWordMatches(matrix, point, "XMAS");
            }
        }

        return occurrences;
    }

    private static int FindWordMatches(Matrix<char> matrix, Point starting, ReadOnlySpan<char> chars)
    {
        if (matrix.ItemAt(starting) != chars[0])
            return 0;

        ReadOnlySpan<Point> directions =
        [
            Point.FromDirection(Direction.Right),
            Point.FromDirection(Direction.Left),
            Point.FromDirection(Direction.Down),
            Point.FromDirection(Direction.Up),
            Point.FromDirection(Direction.Right | Direction.Down),
            Point.FromDirection(Direction.Right | Direction.Up),
            Point.FromDirection(Direction.Left | Direction.Down),
            Point.FromDirection(Direction.Left | Direction.Up),
        ];

        int matches = 0;

        foreach (var direction in directions)
        {
            var current = starting;
            for (int i = 0; i < chars.Length; i++)
            {
                char? currChar = matrix.ItemAt(current);

                // No match or out of matrix bounds.
                if (currChar != chars[i])
                    break;

                current += direction;

                // A match made in heaven.
                if (i == chars.Length - 1)
                    matches++;
            }
        }

        return matches;
    }
}

File: go/problems/day4.go
package problems

import (
	"image"
	"strings"

	"github.com/Meryovi/hobby-aoc-2024/go/shared"
)

type Day4 struct{}

func (d Day4) Solve(input string) int {
	r := 0

	matrix := strings.Split(input, shared.NewLine)
	my := len(matrix)
	mx := len(matrix[0])

	for x := range mx {
		for y := range my {
			p := image.Point{x, y}
			r += countWordMatches(matrix, "XMAS", p)
		}
	}

	return r
}

func countWordMatches(matrix []string, word string, pos image.Point) int {
	dirs := []image.Point{{1, 0}, {-1, 0}, {0, 1}, {0, -1}, {1, 1}, {1, -1}, {-1, 1}, {-1, -1}}
	matches := 0

	for _, dir := range dirs {
		p := pos
		for i := range word {
			if p.Y >= len(matrix) || p.X >= len(matrix[0]) || p.Y < 0 || p.X < 0 {
				break
			}

			if word[i] != matrix[p.Y][p.X] {
				break
			}

			if i == len(word)-1 {
				matches++
				break
			}

			p = p.Add(dir)
		}
	}

	return matches
}
